Unsafe class discovery

matyama opened the pull request:
### Description
* Adds package `unsafe` for impure code
* Adds module `reflection` with functions for discovery of all implementations of a class

### Test plan
Passing release check and tested in other project from which is the code being migrated.

Commit:
Adds Pipfile and updates setup.py and Makefile accordingly

Commit:
Updates virtual env info in README

Commit:
Adds version to setup requirements and generalizes make targets

Commit:
Adds module reflection with functions for class discovery

matyama commented:
Question: Does this code belong to ftoolz?
Answer: Perhaps not, but I don't really want to start and manage new repo for this kind of stuff :see_no_evil:

tm-drtina commented on ftoolz/unsafe/reflection.py:
```diff
@@ -0,0 +1,157 @@
+    ...         pass
+    >>> abstract(B)
+    False
+    """
```
I would add tests for empty abstract class
```python
    >>> class C(ABC):
    ...     pass
    >>> abstract(C)
    False
```

and for class that inherits from abstract, but does not implement abstract methods
```python
    >>> class D(A):
    ...     pass
    >>> abstract(D)
    True
```

tm-drtina commented on ftoolz/unsafe/reflection.py:
```diff
@@ -0,0 +1,157 @@
+T = TypeVar('T')
+
+
+def abstract(clz: Type[T]) -> bool:
```
I don't think you need `Type[T]`, `Type` should be enough, right? There is no use of the `T` anywhere.

Commit:
Adds couple more rests for predicate abstract

Commit:
Adds small optimization of subclasses

martinbabka commented on ftoolz/unsafe/reflection.py:
```diff
@@ -0,0 +1,168 @@
+    **Warning**: Any side-effect of any import under the package will be
+    executed as a side-effect of this function.
+    """
+    import pkgutil
```
not sure if `pkgutil` is always present

matyama replied:
I'll test it, but it should be..

matyama replied:
should be fine, it's a [std lib](https://docs.python.org/3.6/library/pkgutil.html)

martinbabka reviewed (dismissed):
Except for `pkgutil` requirement, looks perfect.

matyama commented:
Just found out that there are some issues with the import, I'll investigate but lets postpone merging this one :(

Commit:
Merge remote-tracking branch 'origin/master' into ftoolz-unsafe-class-discovery

Commit:
Fixes import of classes from implementations

matyama commented:
Added a fix which solves the import issues while keeping the discovery API (`impementations()`) the same. 

The only difference is that I moved doctest to ordinary UTs because there were some issues with getting the doctest-defined classes from a module object - some doctest thing I guess.
